### PR TITLE
Using precomputed data: preliminaries

### PR DESCRIPTION
**Pull request by micheles**
This is some refactoring needed to implement https://bugs.launchpad.net/oq-engine/+bug/1448585.
A lot more work will follow. The tests are green: https://ci.openquake.org/job/zdevel_oq-risklib/543/

### CODE REVIEW

File: openquake/commonlib/calculators/scenario_damage.py [hunk 1]
```diff
@@ -71,12 +71,16 @@ class ScenarioDamageCalculator(base.RiskCalculator):
...
+    pre_calculator = 'scenario'
     core_func = scenario_damage
     result_kind = 'damages_by_key'
 
     def pre_execute(self):
+        if 'gmfs' in self.oqparam.inputs:
+            self.pre_calculator = None
         base.RiskCalculator.pre_execute(self)
+        if 'gmfs' in self.oqparam.inputs:
+            self.precalc.assets_by_site = self.assets_by_site
```
Comment (ptormene) [inline openquake/commonlib/calculators/scenario_damage.py]: Please check if this condition needs to be evaluated twice

File: openquake/commonlib/calculators/scenario_risk.py [hunk 2]
```diff
@@ -104,23 +104,26 @@ class ScenarioRiskCalculator(base.RiskCalculator):
...
         """
         Compute the GMFs, build the epsilons, the riskinputs, and a dictionary
         with the unit of measure, used in the export phase.
         """
+        if 'gmfs' in self.oqparam.inputs:
+            self.pre_calculator = None
         base.RiskCalculator.pre_execute(self)
-        gmfs = base.get_gmfs(self)
+        if 'gmfs' in self.oqparam.inputs:
+            self.precalc.assets_by_site = self.assets_by_site
```
Comment (ptormene) [inline openquake/commonlib/calculators/scenario_risk.py]: Same as above

**Comment by ptormene** [comment]
Apart from the minor comments, LGTM

File: openquake/commonlib/calculators/scenario_damage.py (hunk 1 above)
Comment (micheles) [inline openquake/commonlib/calculators/scenario_damage.py]: No, indeed, it was a leftover, good catch!

**Pull request merged**